### PR TITLE
NVD inference changes for container

### PR DESCRIPTION
**Pull request by gavrishp**
Changes

- run.sh is flexible to initiate the NVD automation on docker and kubernetes. Default choice goes to "VM"
- kubernetes_run.py is used for creating and executing commands using python client

### CODE REVIEW

File: inference/code/torchserve/utils/system_utils.py [hunk 1]
```diff
@@ -63,4 +63,11 @@ def try_and_handle(cmd, dry_run=False):
...
         print(f"Filepath does not exist - {filepath}")
-        sys.exit(1)
\ No newline at end of file
+        sys.exit(1)
+
+def create_folder_if_not_exits(path):
+    isExist = os.path.exists(path)
+    if not isExist:
+        #  Create a new directory because it does not exist
+        os.makedirs(path)
```
Comment (johnugeorge) [inline inference/code/torchserve/utils/system_utils.py:72]: You can use exist_ok flag in os.makedirs

File: inference/code/torchserve/run.sh [hunk 2]
```diff
@@ -36,55 +42,178 @@ do
...
+        mem=${res_array[i+1]}
+    fi
+    done
 
-if [ ! -z $mar_file_path ] ; then
-    cmd+=" --mar $mar_file_path"
-fi
+    if [ -z "$image"  ]
+    then
+        echo "Some or all of the parameters are empty";
```
Comment (johnugeorge) [inline inference/code/torchserve/run.sh:63]: Fix the error message

File: inference/code/torchserve/run.sh [hunk 3]
```diff
@@ -36,55 +42,178 @@ do
...
+        helpFunction
+    fi
 
-if [ ! -z $model_file_path ] ; then
-    cmd+=" --model_path $model_file_path"
-fi
+    if [ -z "$mem"  ] ||  [ -z "$cpu"  ]
```
Comment (johnugeorge) [inline inference/code/torchserve/run.sh:67]: Is this a OR condition? Fix error message here as well

File: inference/code/torchserve/run.sh [hunk 4]
```diff
@@ -36,55 +42,178 @@ do
...
 
-if [ ! -z $model_arch_path ] ; then
-    cmd+=" --model_arch_path $model_arch_path"
-fi
+function create_execution_cmd()
+{
+    gen_folder="gen"
+    if [ $compute_choice = "vm" ] ;
+    then
+        cmd="python3 $wdir/torchserve_run.py"
```
Comment (johnugeorge) [inline inference/code/torchserve/run.sh:79]: This condition check is bit confusing. What is this about?

File: inference/code/torchserve/run.sh [hunk 5]
```diff
@@ -36,55 +42,178 @@ do
...
+
+    if [ -z "$gpus"  ] || [ "$gpus" -eq 0 ] 
+    then
+        gpus="0"
+    fi
+
+    validate_and_get_resources
+
+    mem+='Gi'
+    pip install kubernetes
```
Comment (johnugeorge) [inline inference/code/torchserve/run.sh:153]: Add it in the requirements file? Add client version as well

Comment (gavrishp) [reply inference/code/torchserve/run.sh:153]: This is required only for kubernetes option. Hence had put it here. Redundant for other 2. Any other approach than in common requirements file?

File: inference/code/torchserve/run.sh (hunk 4 above)
Comment (gavrishp) [reply inference/code/torchserve/run.sh:79]: The filepath is not the same in containers and existing approach using current directory cannot be used.

File: inference/code/torchserve/run.sh (hunk 3 above)
Comment (gavrishp) [reply inference/code/torchserve/run.sh:67]: Yes it an OR condition

File: inference/code/torchserve/utils/system_utils.py (hunk 1 above)
Comment (gavrishp) [reply inference/code/torchserve/utils/system_utils.py]: fixed

File: inference/code/torchserve/run.sh (hunk 2 above)
Comment (gavrishp) [reply inference/code/torchserve/run.sh]: Updated

File: inference/code/torchserve/run.sh [hunk 6]
```diff
@@ -36,55 +42,178 @@ do
...
+    fi
+
+    if [ -z "$gpus"  ] || [ "$gpus" -eq 0 ] 
+    then
+        gpus="0"
+    fi
+
+    validate_and_get_resources
+
+    mem+='Gi'
```
Comment (deepanker13) [inline inference/code/torchserve/run.sh:152]: shouldn't memory unit be taken from user

File: inference/code/torchserve/run.sh [hunk 7]
```diff
@@ -36,55 +42,178 @@ do
...
+        docker_init_cmd+=" --gpus $gpus"
+    fi
+
+    echo "$docker_init_cmd"
+    $docker_init_cmd
+
+    docker_exec_cmd="docker exec -it torchserve $cmd"
+
+    echo "$docker_init_cmd"
+    $docker_init_cmd
```
Comment (deepanker13) [inline inference/code/torchserve/run.sh:178]: docker init cmd is coming twice

File: inference/code/torchserve/run.sh [hunk 8]
```diff
@@ -36,55 +42,178 @@ do
...
+
+    if [ -z $stop_server ] ; then
+        echo "Stop Docker Container";
+        docker stop torchserve
+        echo "Remove Docker Container";
+        docker rm torchserve
+    fi
```
Comment (deepanker13) [inline inference/code/torchserve/run.sh:185]: why is stopping server a parameter when we want to execute the container, didn't get it

Comment (deepanker13) [reply inference/code/torchserve/run.sh:185]: so shouldn't stopping be in different function, docker_init_cmd and docker_exec_cmd will start the containers before killing, that I am not getting

File: inference/code/torchserve/run.sh (hunk 6 above)
Comment (gavrishp) [reply inference/code/torchserve/run.sh:152]: This is by design, we should not assume that the User knows the right unit syntax (Gi or GB or g). As in this case it is responsibility of the service to insert the right syntax and define constraints like in current scenario we are asking them to add units in GB 

`
echo -e "\t-r resorces to be provided to the container in quotes '--cpu <cpu to be used> --mem <memory in GBs to be provided to the container>'. Ex: -r '--cpu 16 --mem 24'"`

**Review by johnugeorge — APPROVED** [review]

**Pull request merged**